### ISSUE TITLE
3.1.3 - Mail not working

### ISSUE THREAD

**Issue by rohit1290** [Elgg/Elgg#12835]
Whenever I try to use `elgg_send_email` the mails are not getting delivered and I am getting the following error in my error_log:

```
[23-Sep-2019 22:12:40 Asia/Kolkata] [2019-09-23 22:12:40] ELGG.ERROR: Unable to send mail: Unknown error {"backtrace":[
"[#10] /home1/xxxx/public_html/vendor/elgg/elgg/engine/lib/notification.php:553",
"[#9] /home1/xxxx/public_html/mod/my_external_pages/actions/send_mail.php:60",
"[#8] /home1/xxxx/public_html/vendor/elgg/elgg/engine/classes/Elgg/Router.php:349",
"[#7] /home1/xxxx/public_html/vendor/elgg/elgg/engine/classes/Elgg/Router.php:280",
"[#6] /home1/xxxx/public_html/vendor/elgg/elgg/engine/classes/Elgg/Router.php:130",
"[#5] /home1/xxxx/public_html/vendor/elgg/elgg/engine/classes/Elgg/Router.php:106",
"[#4] /home1/xxxx/public_html/vendor/elgg/elgg/engine/classes/Elgg/Application.php:464",
"[#3] /home1/xxxx/public_html/vendor/elgg/elgg/engine/classes/Elgg/Application.php:359",
"[#2] /home1/xxxx/public_html/vendor/elgg/elgg/engine/classes/Elgg/Application.php:415",
"[#1] /home1/xxxx/public_html/index.php:8"]
```

**Comment by rohit1290** [comment]
And i forgot to mention.. my php mail() is working fine.

**Comment by jdalsem** [comment]
is it only the elgg_send_mail from your `my_external_pages` that is not being sent? Or is it any notification from the site? was it working in 3.1.2?

can you share the code from the send_mail action file?

**Comment by rohit1290** [comment]
code from the send_mail action is a simple contact form.. it get the name, email and message and snd it to the admin email address.

The issue is with all notification.. spam login filter mails, new user registration mail etc.

**Comment by jdalsem** [comment]
did it change with the recent update of Elgg? The error messages comes from an internal function of the mail library, so it points towards a server configuration issue..

**Comment by RiverVanRain** [comment]
Just tested with:
```
if (elgg_send_email('sender@mail', 'to@mail', 'test', 'test')) {
	return elgg_ok_response('', elgg_echo('OK'));
} else {
	return elgg_error_response(elgg_echo('Error'));
}
```
All works very well.

**Comment by rohit1290** [comment]
Now this is interesting..

Here's is my contact page code:
```
use Elgg\Email;

$site = elgg_get_site_entity();
$from = \Elgg\Email\Address::getFormattedEmailAddress($site->getEmailAddress(), $site->getDisplayName());

$emails = explode(',', 'mail_id_1@domain.com, mail_id_2@gmail.com');

$msg = "NAME: "  . get_input('name') ."<br>\n";
$msg .= "EMAIL: "  . get_input('email') ."<br>\n";
$msg .= "MESSAGE: " . nl2br(get_input('comments')) ."<br>\n";

foreach ($emails as $email_address) {
	$email = Email::factory([
		'to' => $email_address,
		'from' => $from,
		'subject' => $site->getDisplayName() . " - Contact Form",
		'body' => $msg,
	]);
	
	$flag = elgg_send_email($email);
}

if ($flag == false) {
	register_error("Error in sending mail!");
} else {
	elgg_clear_sticky_form('contact');
	system_message("Thank you for contacting us!");
}
forward(REFERRER);
```
If I use the above code in a separate page using `\Elgg\Application::start();` the mail are getting delivered properly. The code is like this:

```
$autoload_root = dirname(__DIR__);
if (!is_file("$autoload_root/vendor/autoload.php")) {
	$autoload_root = dirname($autoload_root);
}
require_once "$autoload_root/vendor/autoload.php";

\Elgg\Application::start();

// REST OF THE CONTACT PAGE CODE ABOVE

```

But when I use the exact code as action page I am getting an error.

@RiverVanRain - Can you please confirm if elgg_send_email is working if you call this from an action url?

**Comment by RiverVanRain** [comment]
@rohit1290 This is from an action; contacts form, too

**Comment by rohit1290** [comment]
i will do a fresh local install of 3.1.3 tomorrow and check again.

**Comment by RiverVanRain** [comment]
@rohit1290 Your code works also if you'll fix this error:
```
'to' => _$email_address_,
'from' => **$from**,
```

**Comment by RiverVanRain** [comment]
I did it again on the production website with this action:

```
use Elgg\Email;
use Elgg\Request;

class ContactAction {

	public function __invoke(Request $request) {

		$sender_name = $request->getParam('sender_name');
		$sender_email = $request->getParam('sender_email');
		$email_message = $request->getParam('email_message');
		
		$to_email = elgg_get_site_entity()->email;

		$subject = elgg_echo('contact:subject');

		$message = elgg_echo('contact:notify', [
			$sender_email,
			$sender_name,
			$email_message,
		]);

		if (empty($sender_name) || empty($sender_email) || empty($email_message)) {
			return elgg_error_response(elgg_echo('contact:missingfields'));
		}
		 
		if (!is_email_address($sender_email)) {
			return elgg_error_response(elgg_echo('contact:wrongemail'));
		}
		
		$email = Email::factory([
			'to' => $to_email,
			'from' => $sender_email,
			'subject' => $subject,
			'body' => $message,
		]);
		
		if(elgg_send_email($email)){
			return elgg_ok_response($config, elgg_echo('contact:success'));
		}
		
		return elgg_error_response(elgg_echo('contact:error'));
	}

}
```

And (again) I haven't any issues.

I suggest **closing this  issue**.

**Comment by rohit1290** [comment]
Ok.. i didnt got a chance to test on a fresh install. What i will do is.. I am closing the issue for now.. Let me debug it on my end.. and in case i find something that needs fixing in the core.. i will reopen it..

**Comment by rohit1290** [comment]
I have finally found the problem. It was indeed a server configuration issue. 
sendmail_path was not configured properly and hence the MTA was rejecting the mails.

Fix: Added the following line in php.ini file
`sendmail_path = "/usr/sbin/sendmail -t -i"`

I know the ticket is closed but just wanted to post this here in case any one faces a similar issue in the future then probable this might come handy for them.
